storage-api/src/java/org/apache/hive/common/util/BloomKFilter.java

```diff
@@ -527,20 +527,19 @@ public void add(ElementWrapper wrapper) {
     @Override
     public void run() {
       while (!executor.isTerminated() && !queue.isEmpty()) {
```

Comment:
A bit unrelated, but since you're touching this code.  This check is completely useless:

```
while (!executor.isTerminated() && !queue.isEmpty()) {
  ...
}
```

I cannot think of many scenarios where the thread needs to check the state of its own `ExecutorService`.  If the `ExecutorService` is terminated, it will Interrupt every thread in the pool and that should cause it to cease to run.  Also, checking if the `Queue` is empty is improper.  You will have two threads that check the state of the Queue (size = 1), see the same non-empty queue, and both try to read, even if there is only one item left.  Both should just try to `take` and one will succeed and the other will fail.

Reply:
thanks, your comment regarding isTerminated check seems valid, let me think that over

queue.isEmpty() check is needed I think, please a take a look at the implementation above, every thread uses its own queue (which is filled in advance), there is no concurrency between them, so queue.isEmpty() is a simple check for every thread that "whether I have further things to do, or I can quit",  this way every thread will quit once they're ready, so executor.awaitTermination will return and we're done